binderhub/build.py

```diff
@@ -213,72 +149,137 @@ def get_cmd(self):
-        now = datetime.datetime.now(tz=datetime.timezone.utc)
-        start_cutoff = now - datetime.timedelta(seconds=max_age)
-        deleted = 0
-        for build in builds:
-            phase = build.status.phase
-            phases[phase] += 1
-            annotations = build.metadata.annotations or {}
-            repo = annotations.get("binder-repo", "unknown")
-            delete = False
-            if build.status.phase in {"Failed", "Succeeded", "Evicted"}:
-                # log Deleting Failed build build-image-...
-                # print(build.metadata)
-                app_log.info(
-                    "Deleting %s build %s (repo=%s)",
-                    build.status.phase,
-                    build.metadata.name,
-                    repo,
-                )
-                delete = True
-            else:
-                # check age
-                started = build.status.start_time
-                if max_age and started and started < start_cutoff:
-                    app_log.info(
-                        "Deleting long-running build %s (repo=%s)",
-                        build.metadata.name,
-                        repo,
-                    )
-                    delete = True
-
-            if delete:
-                deleted += 1
-                try:
-                    kube.delete_namespaced_pod(
-                        name=build.metadata.name,
-                        namespace=namespace,
-                        body=client.V1DeleteOptions(grace_period_seconds=0),
-                    )
-                except client.rest.ApiException as e:
-                    if e.status == 404:
-                        # Is ok, someone else has already deleted it
-                        pass
-                    else:
-                        raise
-
-        if deleted:
-            app_log.info("Deleted %i/%i build pods", deleted, len(builds))
-        app_log.debug(
-            "Build phase summary: %s", json.dumps(phases, sort_keys=True, indent=1)
-        )
-
     def progress(self, kind: ProgressEvent.Kind, payload: str):
         """
         Put current progress info into the queue on the main thread
         """
         self.main_loop.add_callback(self.q.put, ProgressEvent(kind, payload))
 
+    def submit(self):
+        """
+        Run a build to create the image for the repository.
+
+        Progress of the build can be monitored by listening for items in
+        the Queue passed to the constructor as `q`.
+        """
+        raise NotImplementedError()
+
+    def stream_logs(self):
+        """
+        Stream build logs to the queue in self.q
+        """
+        pass
+
+    def cleanup(self):
+        """
+        Stream build logs to the queue in self.q
+        """
+        pass
+
+    def stop(self):
+        """
+        Stop watching progress of build
+
+        TODO: What exactly is the purpose of this method?
+        """
+        self.stop_event.set()
```

Comment:
@minrk introduced `stop` and `stop_event` in #628 with the PR description:

> ensures we aren't filling up the build pool with watchers no longer hooked up to any current requests

Reply:
I've added the explanation to the docstring.